# Applied Data Science Capstone from IBM
## Week3
### Segmenting and Clustering Neighborhoods in Toronto

Author: Carlos A. Evangelista Busso

Import Libraries

In [1]:
import requests
import lxml.html as lh
import pandas as pd

The code below allows us to get the data of the HTML table.

In [2]:
url='https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
#Create a handle, page, to handle the contents of the website
page = requests.get(url)
#Store the contents of the website under doc
doc = lh.fromstring(page.content)
#Parse data that are stored between <tr>..</tr> of HTML
tr_elements = doc.xpath('//tr')

For sanity check, ensure that all the rows have the same width. If not, we probably got something more than just the table.

In [3]:
#Check the length of the first 12 rows
[len(T) for T in tr_elements[:12]]

[3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3]

Looks like all our rows have exactly 3 columns. This means all the data collected on tr_elements are from the table.

### Parse Table Header
Next, let’s parse the first row as our header.

In [4]:
tr_elements = doc.xpath('//tr')

#Create empty list
col=[]
i=0

#For each row, store each first element (header) and an empty list
for t in tr_elements[0]:
    i+=1
    name=t.text_content()
    print('%d:"%s"' %(i,name))
    col.append((name,[]))

1:"Postal code
"
2:"Borough
"
3:"Neighborhood
"


### Creating Pandas DataFrame
Each header is appended to a tuple along with an empty list.

In [5]:
#Since out first row is the header, data is stored on the second row onwards
for j in range(1,len(tr_elements)):
    #T is our j'th row
    T=tr_elements[j]
    
    #If row is not of size 3, the //tr data is not from our table 
    if len(T)!=3:
        break
    
    #i is the index of our column
    i=0
    
    #Iterate through each element of the row
    for t in T.iterchildren():
        data=t.text_content() 
        #Check if row is empty
        if i>0:
        #Convert any numerical value to integers
            try:
                data=int(data)
            except:
                pass
        #Append the data to the empty list of the i'th column
        col[i][1].append(data)
        #Increment i for the next column
        i+=1

Just to be sure, let’s check the length of each column. Ideally, they should all be the same.

In [6]:
[len(C) for (title,C) in col]

[181, 181, 181]

Perfect! This shows that each of our 3 columns has exactly 181 values.

In [7]:
Dict={title:column for (title,column) in col}
df=pd.DataFrame(Dict)

Looking at the top 5 cells on the DataFrame:

In [8]:
df.head()

Postal code\n           Borough\n                Neighborhood\n
0         M1A\n      Not assigned\n                            \n
1         M2A\n      Not assigned\n                            \n
2         M3A\n        North York\n                   Parkwoods\n
3         M4A\n        North York\n            Victoria Village\n
4         M5A\n  Downtown Toronto\n  Regent Park / Harbourfront\n

Check the headers of each column

In [9]:
df.columns

Index(['Postal code\n', 'Borough\n', 'Neighborhood\n'], dtype='object')

Rename the headers

In [10]:
df.rename(columns={'Postal code\n':'PostalCode',
                   'Borough\n':'Borough',
                   'Neighborhood\n':'Neighborhood'},inplace=True)
df.columns

Index(['PostalCode', 'Borough', 'Neighborhood'], dtype='object')

It is right!!! Now we can proceed to the next step clean columns erasing the string '\n'

In [11]:
df['PostalCode'] = df['PostalCode'].map(lambda x: x.rstrip('\n'))
df['Borough'] = df['Borough'].map(lambda x: x.rstrip('\n'))
df['Neighborhood'] = df['Neighborhood'].map(lambda x: x.rstrip('\n'))

In [111]:
df.head()

PostalCode           Borough                Neighborhood
0        M1A      Not assigned                            
1        M2A      Not assigned                            
2        M3A        North York                   Parkwoods
3        M4A        North York            Victoria Village
4        M5A  Downtown Toronto  Regent Park / Harbourfront

we drop the rows with 'Not assigned'

In [12]:
dataframe=df.drop(df[df.Borough == 'Not assigned'].index)
dataframe=dataframe.reset_index(drop=True)
dataframe.head()

PostalCode           Borough                                  Neighborhood
0        M3A        North York                                     Parkwoods
1        M4A        North York                              Victoria Village
2        M5A  Downtown Toronto                    Regent Park / Harbourfront
3        M6A        North York             Lawrence Manor / Lawrence Heights
4        M7A  Downtown Toronto  Queen's Park / Ontario Provincial Government

We check if there is any repeated value in the column: 'PostalCode'

In [14]:
len(dataframe.groupby('PostalCode').count())


104

In [114]:
dataframe.shape

(104, 3)

we replace the string: ' /' with ','

In [115]:
dataframe['Neighborhood']=dataframe['Neighborhood'].str.replace(' /',',')
dataframe.head()

PostalCode           Borough                                 Neighborhood
0        M3A        North York                                    Parkwoods
1        M4A        North York                             Victoria Village
2        M5A  Downtown Toronto                    Regent Park, Harbourfront
3        M6A        North York             Lawrence Manor, Lawrence Heights
4        M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

print column name with missing values

In [116]:
dataframe.columns[dataframe.isnull().any()]

Index([], dtype='object')

In [118]:
dataframe

PostalCode                Borough  \
0          M3A             North York   
1          M4A             North York   
2          M5A       Downtown Toronto   
3          M6A             North York   
4          M7A       Downtown Toronto   
5          M9A              Etobicoke   
6          M1B            Scarborough   
7          M3B             North York   
8          M4B              East York   
9          M5B       Downtown Toronto   
10         M6B             North York   
11         M9B              Etobicoke   
12         M1C            Scarborough   
13         M3C             North York   
14         M4C              East York   
15         M5C       Downtown Toronto   
16         M6C                   York   
17         M9C              Etobicoke   
18         M1E            Scarborough   
19         M4E           East Toronto   
20         M5E       Downtown Toronto   
21         M6E                   York   
22         M1G            Scarborough   
23         M4G              East York   
24         M5G       Downtown Toronto   
25         M6G       Downtown Toronto   
26         M1H            Scarborough   
27         M2H             North York   
28         M3H             North York   
29         M4H              East York   
..         ...                    ...   
74         M5R        Central Toronto   
75         M6R           West Toronto   
76         M7R            Mississauga   
77         M9R              Etobicoke   
78         M1S            Scarborough   
79         M4S        Central Toronto   
80         M5S       Downtown Toronto   
81         M6S           West Toronto   
82         M1T            Scarborough   
83         M4T        Central Toronto   
84         M5T       Downtown Toronto   
85         M1V            Scarborough   
86         M4V        Central Toronto   
87         M5V       Downtown Toronto   
88         M8V              Etobicoke   
89         M9V              Etobicoke   
90         M1W            Scarborough   
91         M4W       Downtown Toronto   
92         M5W       Downtown Toronto   
93         M8W              Etobicoke   
94         M9W              Etobicoke   
95         M1X            Scarborough   
96         M4X       Downtown Toronto   
97         M5X       Downtown Toronto   
98         M8X              Etobicoke   
99         M4Y       Downtown Toronto   
100        M7Y           East Toronto   
101        M8Y              Etobicoke   
102        M8Z              Etobicoke   
103             Canadian postal codes   

                                          Neighborhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Regent Park, Harbourfront  
3                     Lawrence Manor, Lawrence Heights  
4          Queen's Park, Ontario Provincial Government  
5                                     Islington Avenue  
6                                       Malvern, Rouge  
7                                            Don Mills  
8                      Parkview Hill, Woodbine Gardens  
9                             Garden District, Ryerson  
10                                           Glencairn  
11   West Deane Park, Princess Gardens, Martin Grov...  
12              Rouge Hill, Port Union, Highland Creek  
13                                           Don Mills  
14                                    Woodbine Heights  
15                                      St. James Town  
16                                  Humewood-Cedarvale  
17   Eringate, Bloordale Gardens, Old Burnhamthorpe...  
18                   Guildwood, Morningside, West Hill  
19                                         The Beaches  
20                                         Berczy Park  
21                                 Caledonia-Fairbanks  
22                                              Woburn  
23                                             Leaside  
24                                  

Drop the last row

In [16]:
dataframe=dataframe.drop(dataframe[dataframe.Borough == 'Canadian postal codes'].index)
dataframe

PostalCode           Borough  \
0          M3A        North York   
1          M4A        North York   
2          M5A  Downtown Toronto   
3          M6A        North York   
4          M7A  Downtown Toronto   
5          M9A         Etobicoke   
6          M1B       Scarborough   
7          M3B        North York   
8          M4B         East York   
9          M5B  Downtown Toronto   
10         M6B        North York   
11         M9B         Etobicoke   
12         M1C       Scarborough   
13         M3C        North York   
14         M4C         East York   
15         M5C  Downtown Toronto   
16         M6C              York   
17         M9C         Etobicoke   
18         M1E       Scarborough   
19         M4E      East Toronto   
20         M5E  Downtown Toronto   
21         M6E              York   
22         M1G       Scarborough   
23         M4G         East York   
24         M5G  Downtown Toronto   
25         M6G  Downtown Toronto   
26         M1H       Scarborough   
27         M2H        North York   
28         M3H        North York   
29         M4H         East York   
..         ...               ...   
73         M4R   Central Toronto   
74         M5R   Central Toronto   
75         M6R      West Toronto   
76         M7R       Mississauga   
77         M9R         Etobicoke   
78         M1S       Scarborough   
79         M4S   Central Toronto   
80         M5S  Downtown Toronto   
81         M6S      West Toronto   
82         M1T       Scarborough   
83         M4T   Central Toronto   
84         M5T  Downtown Toronto   
85         M1V       Scarborough   
86         M4V   Central Toronto   
87         M5V  Downtown Toronto   
88         M8V         Etobicoke   
89         M9V         Etobicoke   
90         M1W       Scarborough   
91         M4W  Downtown Toronto   
92         M5W  Downtown Toronto   
93         M8W         Etobicoke   
94         M9W         Etobicoke   
95         M1X       Scarborough   
96         M4X  Downtown Toronto   
97         M5X  Downtown Toronto   
98         M8X         Etobicoke   
99         M4Y  Downtown Toronto   
100        M7Y      East Toronto   
101        M8Y         Etobicoke   
102        M8Z         Etobicoke   

                                          Neighborhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                           Regent Park / Harbourfront  
3                    Lawrence Manor / Lawrence Heights  
4         Queen's Park / Ontario Provincial Government  
5                                     Islington Avenue  
6                                      Malvern / Rouge  
7                                            Don Mills  
8                     Parkview Hill / Woodbine Gardens  
9                             Garden District, Ryerson  
10                                           Glencairn  
11   West Deane Park / Princess Gardens / Martin Gr...  
12            Rouge Hill / Port Union / Highland Creek  
13                                           Don Mills  
14                                    Woodbine Heights  
15                                      St. James Town  
16                                  Humewood-Cedarvale  
17   Eringate / Bloordale Gardens / Old Burnhamthor...  
18                 Guildwood / Morningside / West Hill  
19                                         The Beaches  
20                                         Berczy Park  
21                                 Caledonia-Fairbanks  
22                                              Woburn  
23                                             Leaside  
24                                  Central Bay Street  
25                                            Christie  
26                                           Cedarbrae  
27                                   Hillcrest Village  
28   Bathurst Manor / Wilson Heights / Downsview North  
29                                    Thorncliffe Park  
..  

Shape of dataframe

In [20]:
dataframe.shape

(103, 3)